File: src/main/java/com/ljs/ootp/ai/selection/bench/BenchScorer.java
package com.ljs.ootp.ai.selection.bench;

import com.ljs.ootp.ai.player.Player;
import com.ljs.ootp.ai.player.ratings.Position;
import com.ljs.ootp.ai.selection.lineup.Lineup;
import com.ljs.ootp.ai.stats.BattingStats;
import com.ljs.ootp.ai.stats.TeamStats;

import java.util.ArrayList;
import java.util.List;

import com.google.common.base.Function;
import com.google.common.base.Predicate;
import com.google.common.base.Predicates;
import com.google.common.collect.ImmutableList;
import com.google.common.collect.Iterables;
import com.google.common.collect.Ordering;

/**
 *
 * @author lstephen
 */
public class BenchScorer {

    private final TeamStats<BattingStats> predictions;

    public BenchScorer(TeamStats<BattingStats> predictions) {
        this.predictions = predictions;
    }

    public Double score(Player toScore, Iterable<Player> bench, Lineup lineup, Lineup.VsHand vs) {
        return score(bench, lineup, vs, Predicates.equalTo(toScore));
    }

    public Double score(Iterable<Player> bench, Lineup lineup, Lineup.VsHand vs) {
        return score(bench, lineup, vs, Predicates.<Player>alwaysTrue());
    }

    private Double score(Iterable<Player> bench, Lineup lineup, Lineup.VsHand vs, Predicate<Player> predicate) {
        Double score = 0.0;

        for (Lineup.Entry entry : lineup) {
            if (!entry.getPositionEnum().equals(Position.PITCHER)) {
                Integer count = 0;
                for (Player p : Iterables.limit(selectBenchPlayer(bench, lineup, vs, entry.getPositionEnum()), 2)) {
                    count++;
                    if (predicate.apply(p)) {
                        Double countFactor = 1.0;
                        Double positionFactor = getPositionFactor(entry.getPositionEnum());

                        if (count > 1) {
                            countFactor = 0.2;
                            positionFactor = .04;
                        }

                        score += (positionFactor * countFactor * vs.getStats(predictions, p).getWobaPlus());
                        score +=
                            (positionFactor * countFactor * p.getDefensiveRatings().getPositionScore(entry.getPositionEnum())) / 2;
                    }
                }
            }
        }

        return score;
    }

    private Double getPositionFactor(Position pos) {
        switch (pos) {
            case CATCHER:
            case SHORTSTOP:
                return .2;
            case THIRD_BASE:
            case SECOND_BASE:
                return .16;
            case CENTER_FIELD:
                return .12;
            case LEFT_FIELD:
            case RIGHT_FIELD:
                return .08;
            default:
                return .04;
        }
    }

    private ImmutableList<Player> selectBenchPlayer(Iterable<Player> bench, Lineup lineup, final Lineup.VsHand vs, Position pos) {
        ImmutableList<Player> sortedBench = Ordering
            .natural()
            .reverse()
            .onResultOf(new Function<Player, Integer>() {
                public Integer apply(Player p) {
                    return vs.getStats(predictions, p).getWobaPlus();
                }
            })
            .compound(Player.byTieBreak())
            .immutableSortedCopy(bench);

        List<Player> ps = new ArrayList<>();

        for (Player p : sortedBench) {
            if (p.canPlay(pos)) {
                ps.add(p);
            }
        }
        return ImmutableList.copyOf(ps);
    }

}

File: src/main/scala/com/github/lstephen/ootp/ai/selection/bench/BenchScorer.scala
package com.github.lstephen.ootp.ai.selection.bench

import com.github.lstephen.ootp.ai.player.Player
import com.github.lstephen.ootp.ai.player.ratings.Position
import com.github.lstephen.ootp.ai.regression.Predictor
import com.github.lstephen.ootp.ai.score.Score
import com.github.lstephen.ootp.ai.selection.depthchart.DepthChart.Backup
import com.github.lstephen.ootp.ai.selection.depthchart.DepthChartSelection
import com.github.lstephen.ootp.ai.selection.lineup.Defense
import com.github.lstephen.ootp.ai.selection.lineup.InLineupScore
import com.github.lstephen.ootp.ai.selection.lineup.Lineup
import com.github.lstephen.ootp.ai.selection.lineup.Lineup.VsHand

import collection.JavaConversions._

class BenchScorer(implicit predictor: Predictor) {
  val depthChartSelection = new DepthChartSelection

  def score(bench: java.lang.Iterable[Player],
            lineup: Lineup,
            vs: VsHand): Double = {
    score_(bench, lineup, vs).toDouble
  }

  def score_(bench: Traversable[Player], lineup: Lineup, vs: VsHand): Score = {
    val dc = depthChartSelection.select(lineup, bench.toSet, vs)

    lineup
      .filter(_.getPositionEnum != Position.PITCHER)
      .flatMap(bu => dc.getBackups(bu.getPositionEnum))
      .map(score(_, vs))
      .total
  }

  def score(bu: Backup, vs: VsHand): Score =
    (Defense.getPositionFactor(bu.getPosition) * bu.getPercentage / 100.0) *: InLineupScore(
      bu.getPlayer,
      bu.getPosition,
      vs).score

}
